Test branch with different slider behaviors

juzen2003 opened the pull request:
- This is a test branch for us to experiment different slider behaviors in gallery & table view.
- This branch is based on: https://github.com/SETI/pds-opus/pull/811
- Local branch: test_branch_with_different_slider_behavior_based_on_slider_issue_807_and_infiniteScroll_issue_761
- Current behavior:
    - Example 1: (13 images per row in gallery view)
        - Go to gallery view.
        - Move slider to a number (7190), it will match startObs (7190) in gallery view.
        - Switch to table view, slider number will match the one in gallery view (7190).
        - In table view, scroll the scrollbar so that 7198 is the top item (slider will move from 7190 to 7198 with step size 1).
        - Switch to gallery view, slider will move back to row boundary 7190.
        - Switch to table view, slider will move back to 7198.
        - In table view, scroll the scrollbar so that infiniteScroll load is triggered and 7283 is the top item (slider is 7283).
        - Switch to gallery view, the slider will move to new row boundary 7281.
        - Switch to table view, the slider will move back to 7283.
        - Switch to gallery view, and move the slider from 7281 to 13040.
        - Switch to table view, the slider will be 13040.

    - Example 2: (13 images per row in gallery view)
        - Go to table view.
        - Move slider to 3867, top item will be obs 3867.
        - Switch to gallery view, slider will move to 3862 (row boundary). Obs 3867 is one of item in top row.
        - Switch to table view, slider will move to 3867.
        - Switch to gallery view and scroll up scrollbar so that infiniteScrolll load is triggered, and slider move to 3797.
        - Switch to table view, slider is at 3797.

juzen2003 committed:
- In browse.js (line 612, 623-624, 634), update setScrollbarPosition
  function to take in one more parameter "offset" (default to 0), this
  will be used to more properly set the scrollbar location so that
  scrolling looks smoother.
- In browse.js (line 757-773, 781-782, 786-787), update updateSliderHandle
  to get the scrollbar offset and store it in infiniteScroll instances as
  "scrollbarOffset". This will be used later on to pass in
  setScrollbarPosition after infiniteScroll load event is triggered and
  renderGalleryAndTable is called.
- In browse.js (line 656-657, 661, 666), update onUpdateSlider to set
  offset to 0 in infiniteScroll instances so that the startObs item can be
  displayed fully after slider has done moving.
- In browse.js (line 1569-1572), update path in infiniteScroll. We will
  update the obsNum in InfiniteScroll instances with the startObs values
  right before infiniteScroll load event is triggered. This will be used
  to properly set the scrollbar and slider location after new data is loaded.
- In browse.js (line 1323), pass in infiniteScroll instances' scrollbarOffset
  to setScrollbarPosition. This will make scrolling looks smoother when
  infiniteScroll load event is triggered.

juzen2003 committed:
- In browse.js (line 714-721), update updateSliderHandle so that when
  calculating startObs in table view, we are using the 2nd table tr's
  height. This is because in some corner cases, the 1st table tr's height
  is 1px larger than rest of tr and it will mess up the calculation.

juzen2003 committed:
- In browse.js (line 753), in table view, we divide by the 2nd table tr's
  height because in some corner cases, the first table tr's height will
  be 1px larger than rest of tr, and this will mess up the calculation.
- In browse.js (line 699-711, 728-854), break down steps in updateSliderHandle
  into multiple functions and add comments.

juzen2003 committed:
- In browse.js (line 1190-1191), update updateBrowseNav function to set
  the scrollbar position using scrollbarObsNum stored in infiniteScroll
  instance in table view.
- In browse.js (line 773-786), modify getCurrentStartObs function to properly
  update currentScrollObsNum so that later on table view scrollbar location
  stays at where it's been left off when we switch back to table view again.

juzen2003 committed:
- In browse.js (line 727), rename "getCurrentStartObs" function to
  "realignDOMAndGetStartObsAndScrollbarObsNum".
- In browse.js (line 743), rename variable "calculatedFirstObs" to
  "alignedCachedFirstObs".
- In browse.js (line 738-740), move "deleteCachedObs" function call inside
  "realignDOMAndGetStartObsAndScrollbarObsNum".
- In browse.js (line 769-771, 774,777), update the comments for
  currentScrollObsNum and obsNum inside "realignDOMAndGetStartObsAndScrollbarObsNum".

juzen2003 committed:
- In browse.js (line 781-795), modify the logic about getting obsNum and
  currentScrollObsNum. Now in gallery view, if scrollbarObsNum in
  infiniteScroll instance is:
  (1) still within the current startObs' boundary
  (2) larger than max slider value when current startObs is equal to
      max slider value
  We want to make sure it didn't get updated to startObs. This will make
  sure table view scrollbar location stays at where it's been left off
  when we switch back to table view again.

juzen2003 committed:
- In browse.js (line 790), adding ;

juzen2003 committed:
Merge branch 'infiniteScroll_issue_761' into slider_issue_807

juzen2003 committed:
- In browse.js (line 699-700), rename "numThumbnails" to "numObservations".
- In browse.js (line 807), rename "deleteCachedObs" to
  "deleteObsToCorrectRowBoundary".
- In browse.js (line 844-848), update updateSliderValue function to take
  less parameters, instead of taking in those parameters, those parameters
  are obtained inside the function.
- In browse.js (line 896-902), fix the indentation.
- In browse.js (line 1305, 1407-1408), simplify the code a little bit by
  making infiniteScrollDataObj = $(selector).data("infiniteScroll").options;

juzen2003 committed:
- In browse.js (line 807, 908), update the return of
  realignDOMAndGetStartObsAndScrollbarObsNum and getScrollbarOffset. Now
  these two functions will return an object instead of a array.

juzen2003 committed:
- In browse.js (line 735, 810, 830, 892, ), modify
  realignDOMAndGetStartObsAndScrollbarObsNum, deleteObsToCorrectRowBoundary,
  getScrollbarOffset and getSliderMaxValue to take in less parameters.
  Parameters like: tab, viewNamespace, and galleryBoundingRect will be
  obtained in each function.

juzen2003 committed:
- In browse.js (line 757-760), declare browseContentsContainerTop to store
  $(`${tab} .gallery-contents`).offset().top to make the code cleaner.
- In browse.js (716-717), remove declaration of galleryOffset and tableOffset.
  Since they are used for one time only, we pass in the values from the
  object directly.
- In browse.js (line 1664), remove unnecessary line
  scrollbarObsNum = Math.max(scrollbarObsNum, 1);

juzen2003 committed:
- In browse.js (line 705-710), update comments in updateSliderHandle.
- In browse.js (line 740-750), update comments in
  realignDOMAndGetStartObsAndScrollbarObsNum.
- In browse.js (line 823-831), update comments in
  deleteObsToCorrectRowBoundary.

juzen2003 committed:
- In browse.js (line 760-762), modify realignDOMAndGetStartObsAndScrollbarObsNum
  to update firstCachedObs after browser is resized. This will make sure
  slider is updated to the correct value when browser is resized. Previously
  we didn't update firstCachedObs after browser is resized, so the obsNum
  is one row ahead at first. And the correct obsNum gets updated when the
  mutationObserver detects changes in gallery view. That's why we saw the
  slider changes the value twice when window is resized. Now slider will
  change to the correct value right away.

juzen2003 committed:
- Commit before switching branches, leave debugging console.log in browse.js.

juzen2003 committed:
- In browse.js (line 821-835), update the logic in realignDOMAndGetStartObsAndScrollbarObsNum
  so that When resizing happened and some observations are deleted to
  correct the row boundary:
  (1) In table view, if scrollbarObsNum is still within the current startObs'
  boundary, scrollbar will stay at where it is.
  (2) In table view, if scrollbarObsNum is not within the current startObs'
  boundary, scrollbar will be moved to the location where slider value
  (current startObs) is the top item.

juzen2003 committed:
- In browse.js (line 725), update the if condition to (opus.prefs[startObsLabel]
  + numObservations - 1) so that we can get the correct number of rendered
  observations.

juzen2003 committed:
- In browse.js (line 791, 802-803), add defensive programming to make sure
  obsNum is never less than 1.

juzen2003 committed:
- In browse.js (line 640), rename function "onUpdateSliderHandle" to
  "onSliderHandleMoving".
- In browse.js (line 646), rename function "onUpdateSlider" to
  "onSliderHandleStop".
- In browse.js (line 816-853), create a function "calculateCurrentStartObs"
  to include the calculation of obsNum (startObs).

juzen2003 committed:
- In browse.js (line 842-854), update the return of calculateCurrentStartObs
  function.

juzen2003 committed:
- In opus.css (line 110-118), remove min-height for #browse .gallery so
  that there is not extra space for scrolling when #browse .gallery height
  is less than 700px.
- In browse.js (line 738-739), make sure slider is moved to the most left
  when it's disabled.
- In browse.js (line 725-730), update the if condition to make sure when
  slider has no value to scroll, it should always be disabled.
- In browse.js (641, 647, 1693, 1711, 1734), in cart.js (line 249), make
  sure startobs is never less than 1.

juzen2003 committed:
- In browse.js (line 906-907), update getSliderMaxValue to make sure
  return maxSliderVal is never less than 1.

juzen2003 committed:
- Commit before switching branches, all modifications are done in browse.js

juzen2003 committed:
- In browse.js (line 838-839, 873, 920-921), update the calculation for
  startobs and maxSliderVal. This is used to separate the startobs for
  gallery and table view.
- In browse.js (line 963-970), update the slider step size to 1 when it's
  in table view.

juzen2003 committed:
- In browse.js (line 660-676, 702-704, 708-709), modify onSliderHandleStop
  to properly update data in infiniteScroll instances, and also call
  loadData properly when scrolling slider in table view.

rfrenchseti commented on opus/application/static_media/css/opus.css:
```diff
@@ -107,13 +107,14 @@ body {
     background-color: white;
 }
 
+/* Note: this is the scrolling area for thumbnail, when min-height is set
```
I'm not sure I understand the purpose of this comment. Since min-height is no longer set, will anyone care about this?
If you want to leave it, then:
- Say "thumbnails" and "thumbnails are"
- lower case "we"

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -630,36 +631,66 @@ var o_browse = {
             "scrollbarObsNum": value,
+            "scrollbarOffset": 0
         });
+
```
You can go ahead and kill these lines

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -668,11 +699,14 @@ var o_browse = {
             // (one current page, one next page, and one previous page) starting from obsNum.
             // obsNum will be the very first obs for data rendering this time
-            let obsNum = Math.max(value - o_browse.getLimit(), 1);
+            // let obsNum = Math.max(value - o_browse.getLimit(), 1);
```
... and this line

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -668,11 +699,14 @@ var o_browse = {
             // If obsNum is 1, previous page will have value - 1 items, so we render value - 1 + 2 * o_browse.getLimit() items
             // else we render 2 * o_browse.getLimit() items.
-            let customizedLimitNum = obsNum === 1 ? value - 1 + 2 * o_browse.getLimit() : 3 * o_browse.getLimit();
+            // let customizedLimitNum = obsNum === 1 ? value - 1 + 2 * o_browse.getLimit() : 3 * o_browse.getLimit();
```
...and this line

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -691,111 +725,298 @@ var o_browse = {
+            // 1. Get the current startObs (top left item in gallery view in current page)
+            //    and top obsNum (top item in table view and one of top row items in gallery
+            //    view) that scrollbar is at.
+            //    - If browser is resized, realign doms by deleting some cached obSnum to make
```
Capitalize "DOMs" since it's an acronum. "obsNum".

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -691,111 +725,298 @@ var o_browse = {
+            // 2. Get the max value for slider.
+            // 3. Get the scrollbar offset for both gallery and table view.
+            // 4. Update the slider value and data in infiniteScroll instances and URL using
+            //    the data obtained from above 2 steps.
```
3 steps?

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -691,111 +725,298 @@ var o_browse = {
+            let maxSliderVal = o_browse.getSliderMaxValue();
+            let scrollbarOffset = o_browse.getScrollbarOffset(obsNum, currentScrollObsNum);
+            o_browse.setSliderParameters(obsNum, currentScrollObsNum, scrollbarOffset.galleryOffset,
+                                       scrollbarOffset.tableOffset);
```
Indent to align with above arguments

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -691,111 +725,298 @@ var o_browse = {
+            (currentSliderValue <= 1 && currentSliderValue >= currentSliderMax)) {
             // disable the slider because the observations don't fill the browser window
-            // $("#op-observation-slider").slider({
+            // $(`${tab} #op-observation-slider`).slider({
```
You can remove these comments now that I had put in before

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -691,111 +725,298 @@ var o_browse = {
             //     "max": 1,
             // });
             $(`${tab} .op-slider-pointer`).css("width", "3ch");
+            // Make sure slider always move the the most left before being disabled.
```
move "to" the

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -691,111 +725,298 @@ var o_browse = {
             // });
             $(`${tab} .op-slider-pointer`).css("width", "3ch");
+            // Make sure slider always move the the most left before being disabled.
+            $(`${tab} .op-observation-number`).html("1");
```
Does setting the HTML to "1" actually move the slider? I thought you would have to change it's "value" property.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -691,111 +725,298 @@ var o_browse = {
 
+    realignDOMAndGetStartObsAndScrollbarObsNum: function(selector, browserResized) {
+        /**
+         * Get the current startObs (obsNum, for slider number) and the
```
I don't understand this first sentence. Capitalize "DOM" everywhere. Remove period after "resizing".

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -691,111 +725,298 @@ var o_browse = {
+
+    calculateCurrentStartObs: function(selector, firstCachedObs) {
+        /**
+         * Calculation of getting the current startObs based on rendered
```
Remove "getting"

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -691,111 +725,298 @@ var o_browse = {
+        let galleryObsElem = $(`${tab} .gallery [data-obs]`);
+        let tableObsElem = $(`${tab} .op-data-table-view [data-obs]`);
+        // delete first "numToDelete" obs if row size is changed
+        if (numToDelete !== 0) {
```
We don't need the `if` test because the loop won't execute at all if numToDelete is 0.

rfrenchseti reviewed (changes requested):
- Size the browser such that the gallery view has an inactivate slider but the table view doesn't. Go to the table view. Move the slider to "1". The slider in the table view will become inactive and you can't get it back.

- There is still the problem that when pasting a URL with a startobs that is not aligned with the current browser size, the slider is correct but the actual fetched data is wrong. The top left obs in gallery will equal the startobs slug, not the slider, and the whole DOM will be misaligned. Now would be a good time to fix that.

- I can't tell from the code - do we still have the problem where a whole bunch of resizing will eat away at the DOM and eventually screw up what we're actually seeing?

- I thought we were going to make it so when you resize, the top-left obs is always still in the top row. This means the slider number should always decrease. I don't see that behavior when increasing the browser width.

juzen2003 committed:
- In opus.css (line 109), remove unwanted comments.
- In browse.js (line 682, 690, 695), modify onSliderHandleStop, remove
  unwanted comments.
- In browse.js (line 897), removed unnecessary if condition in function
  deleteObsToCorrectRowBoundary.
- In browse.js (line 724, 730, 737, 757-767, 828), modify comments and
  clean up code style.
- In browse.js (line 748-749), properly update slider value to 1 so that
  it will be moved to the most left before being disabled.

juzen2003 committed:
- In browse.js (line 752), update selector to $(`${tab} #op-observation-slider`),
  this will properly move the slider to 1 before slider being disabled.
- In browse.js (line 745-747), update the conditions to properly disable
  slider in gallery & table views.

juzen2003 committed:
- In browse.js (line 745, 747), change "sliderDisabledCondition" to
  "numberOfObsFitOnTheScreen".

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into test_branch_with_different_slider_behavior_based_on_slider_issue_807_and_infiniteScroll_issue_761

juzen2003 committed:
- In browse.js (line 1824-1828), modify loadData to adjust startobs based
  on correct row boundary. This will make sure when URL with a startobs
  not aligned with current browser size is pasted, that startobs will be
  adjusted to a new value which is aligned with the browser size.

juzen2003 committed:
- In mutationObserver.js (line 43), update debounced adjustBrowseHeight
  to check if DOM realignment is needed for switchGalleryAndTableObserver.
  This will make when switching from table to gallery view, DOM realignment
  is performed if needed.
- In browse.js (line 783-789), delete cached obs when we are in gallery
  view or switch from table to gallery view.
- In browse.js (line 796-807), update the logic to calculate obsNum when
  resize happened and some obs are deleted. This will make sure startObs
  before resizing is still in top row of gallery view and also make new
  startObs smaller than previous startObs when resizing happened.
- In browse.js (line 820-826), manually set the scrollbar when resizing
  happened.

juzen2003 committed:
- In browse.js (line 808-822), add the logic to properly update
  currentScrollObsNumin gallery view, if scrollbarObsNum in infiniteScroll
  instance is:
  (1) still within the current startObs' boundary.
  (2) larger than max slider value when current startObs is equal to max
  slider value.
  We want to make sure it didn't get updated to startObs. This will make
  sure table view scrollbar location stays at where it's been left off
  when we switch back to table view again.

juzen2003 committed:
- In browse.js (line 746), fix the indentation.

juzen2003 committed:
- In browse.js (line 654, 687-696), update the logic in onSliderHandleStop.
  Make sure when slider is moved to an obs close to the last cached obs
  (less than the number of items fit on one page), we will load more data.
  This will make sure slider value stays at where it's been dragged to even
  if that value is very close to the edge of cached obs.

rfrenchseti reviewed (changes requested):
- Whenever infinite scroll gets triggered, the gallery realigns with the top edge of the browser, so scrolling isn't really smooth anymore
- If you paste a URL that is in table view, it rounds the startobs down to the next gallery row instead of just leaving it there

juzen2003 committed:
- In browse.js (line 839-845), update the logic to set offset for scrollbar
  position so that it will have smooth scrolling in both gallery and table
  view when infiniteScroll load is triggered.

juzen2003 committed:
- In browse.js (line 1847-1851), modify the way we update startObs in
  loadData. In gallery view, we will round down the startObs based on the
  current browser size. If it's in table view, we won't round down the
  startObs.

juzen2003 committed:
- In browse.js (line 1740-1752, 1758-1771), modify infiniteScroll path
  function. When loading a pasted URL in table view with a startObs
  not aligned with current browser size, we calculate the alignedCachedFirstObs
  and use it as the startObs in the path for infiniteScroll to load previous
  page, so that when load prev page is triggered, all rendered obs will
  be aligned with the browser size.
- In browse.js (line 1918-1924), modify loadData. When pasting a URL,
  prefetch some data from previous page so that scrollbar will stay in the
  middle.
- In browse.js (line 726, 734-736, 739), removed unused variables and update
  comments.

rfrenchseti reviewed (changes requested):
There is a problem with moving the slider all the way to the right end when it doesn't trigger a data load. The easiest way to see this is:
- Go to the gallery
- Move the slider all the way to the right
- Go to the table
- Move the slider all the way to the right (since you can now go a little further)
- Note that the table doesn't actually move
The same thing happens in the gallery if you manage to flush the cache, more the slider almost all the way to the right, and then move it the rest of the way. The gallery won't move either.

juzen2003 committed:
- In browse.js (line 691), update the if condition in onSliderHandleStop.
  When dragging slider, if the cached lastObs is the last obs based on the
  search, we will set the scrollbar location instead of load next page.
  This will make sure when dragging slider all the way to the right and
  there is no data load triggered, scrollbar position will be set properly.

juzen2003 committed:
- In browse.js (line 690-692), update comments in onSliderHandleStop.

juzen2003 committed:
Fix issue of startObs and slider not matching when resizing browser.
- In browse.js (line 837-838), update the offset using the one stored in
  infiniteScroll instances, instead of recalculating again. If we call
  getScrollbarOffset here, the return offset will be off when some cached
  obs are deleted (and before all DOMs are properly aligned).
- In browse.js (line 1739-1740), update infiniteScroll path function. We
  will call countGalleryImages to get the number of items in a row inside
  the path so that alignedCachedFirstObs will be calculated properly
  every time the function is called.

juzen2003 committed:
Fix the issue to have last row fully displayed in browse tab when slider
is moved to the end.
- In browse.js (line 831-845), add the logic to properly set obsNum so that
  the last row is fully displayed in browse tab when scrollbar reaches to
  the end of the data.
- In browse.js (line 891-895) and opus.js (line 105-108), add sliderViewableFraction
  to determine the viewable percentage of an item in browse tab so that it can
  be treated as startobs.
- In browse.js (line 953-956), add one more row to max slider value. This
  will make sure the last row is fully displayed when slider is moved to
  the end.

juzen2003 committed:
Clean up comments.
- In browse.js (line 831-832), update comments.

juzen2003 committed:
Fix the issue of mismatching slider value in table view when pasting an
URL.
- In browse.js (1777-1778), properly update the customizedLimitNum when
  we paste an URL and try to prefetch one previous page starting from
  obsNum. This will make sure the first prefetched page (starting from
  obsNum = 1) is fetching correct number of data.
- In browse.js (1785-1786), modify to properly update data stored in
  infiniteScroll instances and opus.prefs.startobs.

juzen2003 committed:
Fix the jumping slider issue when reloading page.
- In browse.js (line 1995-2000), call adjustBrowseHeight and adjustTableSize
  to to make sure the height of .op-gallery-view and .op-data-table-view
  are set. This will prevent the height of data obs containers from jumping
  when the page is loaded, and avoid the wrong calculation of container
  position in setScrollbarPosition.

juzen2003 committed:
Fix the issue of mismatching slider value and startobs when switching
between gallery & table view after resizing.
- In mutationObserver.js (line 43-50), use the non-debounced version of
  adjustBrowseHeight & adjustTableSize. This will make sure the height of
  .op-gallery-view and .op-data-table-view are set right away when switching
  to browse tab or switching between gallery & table view. That way the
  calculation of setScrollbarPosition will be correct and slider value
  will match startObs.
